File: src/main/python/AoC2019_01.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 1
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST1 = "12"
TEST2 = "14"
TEST3 = "1969"
TEST4 = "100756"

Input = list[int]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return [int(_) for _ in input_data]

    def fuel_for_mass(self, m: int) -> int:
        return m // 3 - 2

    def part_1(self, inputs: Input) -> int:
        return sum(self.fuel_for_mass(m) for m in inputs)

    def rocket_equation(self, m: int) -> int:
        total_fuel = 0
        while (fuel := self.fuel_for_mass(m)) > 0:
            total_fuel += fuel
            m = fuel
        return total_fuel

    def part_2(self, inputs: Input) -> int:
        return sum(self.rocket_equation(m) for m in inputs)

    @aoc_samples(
        (
            ("part_1", TEST1, 2),
            ("part_1", TEST2, 2),
            ("part_1", TEST3, 654),
            ("part_1", TEST4, 33583),
            ("part_2", TEST1, 2),
            ("part_2", TEST3, 966),
            ("part_2", TEST4, 50346),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2019, 1)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2019_01.java
import java.util.Arrays;
import java.util.List;
import java.util.function.IntUnaryOperator;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2019_01 extends SolutionBase<int[], Integer, Integer> {
    
    private AoC2019_01(final boolean debug) {
        super(debug);
    }

    public static AoC2019_01 create() {
        return new AoC2019_01(false);
    }

    public static AoC2019_01 createDebug() {
        return new AoC2019_01(true);
    }
 
    @Override
    protected int[] parseInput(final List<String> inputs) {
        return inputs.stream().mapToInt(Integer::parseInt).toArray();
    }

    private int fuelForMass(final int m) {
        return m / 3 - 2;
    }
    
    private int rocketEquation(final int mass) {
        int totalFuel = 0;
        int fuel = fuelForMass(mass);
        while (fuel > 0) {
            totalFuel += fuel;
            fuel = fuelForMass(fuel);
        }
        return totalFuel;
    }
    
    private int sum(final int[] modules, final IntUnaryOperator strategy) {
        return Arrays.stream(modules).map(strategy).sum();
    }
    
    @Override
    public Integer solvePart1(final int[] modules) {
        return sum(modules, this::fuelForMass);
    }

    @Override
    public Integer solvePart2(final int[] modules) {
        return sum(modules, this::rocketEquation);
    }

    @Samples({
        @Sample(method = "part1", input = "12", expected = "2"),
        @Sample(method = "part1", input = "14", expected = "2"),
        @Sample(method = "part1", input = "1969", expected = "654"),
        @Sample(method = "part1", input = "100756", expected = "33583"),
        @Sample(method = "part2", input = "12", expected = "2"),
        @Sample(method = "part2", input = "1969", expected = "966"),
        @Sample(method = "part2", input = "100756", expected = "50346"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2019_01.create().run();
    }
}
